### PR TITLE
Adds update channel bookmark endpoint

### PR DESCRIPTION
**Pull request by mgdelacroix**
#### Summary
This PR adds the update channel bookmarks endpoint

#### Ticket Link
https://mattermost.atlassian.net/browse/MM-54419

#### Release Note
```release-note
Added update channel bookmark endpoint at /api/v4/channels/{channel_id}/bookmarks/{bookmark_id}.
```

### CODE REVIEW

File: server/channels/api4/channel_bookmark_test.go
Comment (enahum) [inline server/channels/api4/channel_bookmark_test.go:1]: missing test of user A updating the bookmark of user B

File: server/channels/store/sqlstore/channel_bookmark_store.go [hunk 1]
```diff
@@ -157,16 +160,26 @@ func (s *SqlChannelBookmarkStore) Update(bookmark *model.ChannelBookmark) error
...
 	if err != nil {
 		return errors.Wrapf(err, "failed to update channel bookmark with id=%s", bookmark.Id)
 	}
+	rowsAffected, err := res.RowsAffected()
+	if err != nil {
+		return errors.Wrapf(err, "failed to get affected rows after updating bookmark with id=%s", bookmark.Id)
+	}
+	if rowsAffected == 0 {
+		return errors.Wrapf(sql.ErrNoRows, "cannot find bookmark with id=%s for update", bookmark.Id)
+	}
```
Comment (enahum) [inline server/channels/store/sqlstore/channel_bookmark_store.go:182]: 💯

File: server/i18n/en.json
Comment (enahum) [inline server/i18n/en.json:1]: do run `make i18n-extract`

File: server/public/model/channel_bookmark.go [hunk 2]
```diff
@@ -47,11 +47,68 @@ func (o *ChannelBookmark) Auditable() map[string]interface{} {
...
+		"type":    o.Type,
+	}
+}
+
+func (o *ChannelBookmark) Patch(patch *ChannelBookmarkPatch) {
+	if patch.FileId != nil {
+		o.FileId = *patch.FileId
+	}
+
+	if patch.DisplayName != nil {
```
Comment (enahum) [inline server/public/model/channel_bookmark.go:72]: should also be different that empty, no? same with the others or I'm missing something

**Review by enahum — CHANGES_REQUESTED** [review]

File: server/channels/api4/channel_bookmark.go [hunk 3]
```diff
@@ -99,3 +100,99 @@ func createChannelBookmark(c *Context, w http.ResponseWriter, r *http.Request) {
...
+	case model.ChannelTypePrivate:
+		if !c.App.SessionHasPermissionToChannel(c.AppContext, *c.AppContext.Session(), c.Params.ChannelId, model.PermissionEditBookmarkPrivateChannel) {
+			c.SetPermissionError(model.PermissionEditBookmarkPrivateChannel)
+			return
+		}
+
+	case model.ChannelTypeGroup, model.ChannelTypeDirect:
+		// Any member of DM/GMs but guests can manage channel bookmarks
+		if _, errGet := c.App.GetChannelMember(c.AppContext, channel.Id, c.AppContext.Session().UserId); errGet != nil {
+			c.Err = model.NewAppError("createChannelBookmark", "api.channel.bookmark.update_channel_bookmark.direct_or_group_channels.forbidden.app_error", nil, "", http.StatusForbidden)
```
[the remark this answers is not in the document]
Comment (mgdelacroix) [reply server/channels/api4/channel_bookmark.go:163]: Good catch!

File: server/public/model/channel_bookmark.go (hunk 2 above)
Comment (mgdelacroix) [reply server/public/model/channel_bookmark.go:73]: I don't think so. The rationale here is that if the field is present in the patch, we accept its contents, so if we're receiving a `"display_name": ""` that means that the user wants to set its value to empty, even if that's going to cause a validation error down the line

**Review by enahum — APPROVED** [review]

**Pull request merged**